File: tia_bot/mongo/shop.py
"""shop_inventory + shop_daily_stock 어댑터.

shared-db `crud/shop.ts` 시맨틱 그대로 옮김.

- shop_inventory: 사용자(User) 단위 편의점 보유. character_inventory 와 도메인 분리.
- shop_daily_stock: 아이템 일별 재고. lastRefresh 는 KST 'YYYY-MM-DD' 문자열.
  UTC Date 가 아닌 이유 — KST 기준 일자 변경 + 봇 코드와 정합.

Atomic 패턴:
- removeInventory: { quantity: $gte: qty } filter + $inc + race-aware deleteOne.
- reduceStock: { stock: $gte: qty } filter + $inc.

호출자는 user_id_hex (User._id.toHexString()) 만 주입. Discord snowflake 받지 않음.
"""

from datetime import datetime, timezone
from typing import Optional

from pymongo import ReturnDocument
from pymongo.collection import Collection

from .client import get_db

# ─────────────────────────────────────────────
# 컬렉션 이름
# ─────────────────────────────────────────────
_COL_INV = "shop_inventory"
_COL_STOCK = "shop_daily_stock"


def _inv_col() -> Collection:
    return get_db()[_COL_INV]


def _stock_col() -> Collection:
    return get_db()[_COL_STOCK]


# ─────────────────────────────────────────────
# shop_inventory
# ─────────────────────────────────────────────


def get_user_inventory(user_id_hex: str) -> list[dict]:
    """활성 보유만 (quantity > 0). 일반 표시/조회용."""
    if not isinstance(user_id_hex, str) or not user_id_hex:
        raise ValueError(
            f"get_user_inventory: invalid user_id_hex: {user_id_hex!r}"
        )
    return list(
        _inv_col()
        .find({"userId": user_id_hex, "quantity": {"$gt": 0}})
        .sort("itemId", 1)
    )


def get_user_inventory_raw(user_id_hex: str) -> list[dict]:
    """0 quantity 포함 전체 (감사/마이그용)."""
    if not isinstance(user_id_hex, str) or not user_id_hex:
        raise ValueError(
            f"get_user_inventory_raw: invalid user_id_hex: {user_id_hex!r}"
        )
    return list(_inv_col().find({"userId": user_id_hex}).sort("itemId", 1))


def add_inventory(user_id_hex: str, item_id: str, qty: int) -> None:
    """upsert + $inc. qty 양수 강제."""
    if not isinstance(user_id_hex, str) or not user_id_hex:
        raise ValueError(f"add_inventory: invalid user_id_hex: {user_id_hex!r}")
    if not isinstance(item_id, str) or not item_id:
        raise ValueError(f"add_inventory: invalid item_id: {item_id!r}")
    if not isinstance(qty, int) or isinstance(qty, bool):
        raise ValueError(
            f"add_inventory: qty must be int, got {type(qty).__name__}"
        )
    if qty <= 0:
        raise ValueError(f"add_inventory: qty must be positive, got {qty}")

    _inv_col().update_one(
        {"userId": user_id_hex, "itemId": item_id},
        {
            "$inc": {"quantity": qty},
            "$set": {"updatedAt": datetime.now(timezone.utc)},
            "$setOnInsert": {"userId": user_id_hex, "itemId": item_id},
        },
        upsert=True,
    )


def remove_inventory(user_id_hex: str, item_id: str, qty: int) -> bool:
    """atomic 차감.

    - quantity >= qty 일 때만 매치 → race condition 방지.
    - 보유 부족 시 False (수정 없음).
    - 차감 후 quantity == 0 이면 race-aware deleteOne 으로 row 정리.
      (다른 호출이 그 사이 +qty 했으면 quantity:0 매치 안 되어 보존.)

    Returns: 성공 시 True, 보유 부족 시 False.
    """
    if not isinstance(user_id_hex, str) or not user_id_hex:
        raise ValueError(
            f"remove_inventory: invalid user_id_hex: {user_id_hex!r}"
        )
    if not isinstance(item_id, str) or not item_id:
        raise ValueError(f"remove_inventory: invalid item_id: {item_id!r}")
    if not isinstance(qty, int) or isinstance(qty, bool):
        raise ValueError(
            f"remove_inventory: qty must be int, got {type(qty).__name__}"
        )
    if qty <= 0:
        raise ValueError(f"remove_inventory: qty must be positive, got {qty}")

    col = _inv_col()
    result = col.find_one_and_update(
        {"userId": user_id_hex, "itemId": item_id, "quantity": {"$gte": qty}},
        {
            "$inc": {"quantity": -qty},
            "$set": {"updatedAt": datetime.now(timezone.utc)},
        },
        return_document=ReturnDocument.AFTER,
    )
    if result is None:
        return False
    if result.get("quantity") == 0:
        # race-aware: { _id, quantity:0 } filter 자체가 동시 +qty 를 막아준다.
        # delete_one 은 매치 0건이어도 예외를 던지지 않으므로 swallow 불필요.
        col.delete_one({"_id": result["_id"], "quantity": 0})
    return True


# ─────────────────────────────────────────────
# shop_daily_stock
# ─────────────────────────────────────────────


def needs_refresh(item_id: str, today_kst: str) -> bool:
    """lastRefresh 가 today_kst 와 다르면 True. 문서 미존재도 True."""
    if not isinstance(item_id, str) or not item_id:
        raise ValueError(f"needs_refresh: invalid item_id: {item_id!r}")
    if not isinstance(today_kst, str) or not today_kst:
        raise ValueError(f"needs_refresh: invalid today_kst: {today_kst!r}")
    doc = _stock_col().find_one({"itemId": item_id})
    if not doc:
        return True
    return doc.get("lastRefresh") != today_kst


def refresh_stock(item_id: str, stock: int, today_kst: str) -> None:
    """재고를 today_kst 기준으로 강제 갱신 (upsert)."""
    if not isinstance(item_id, str) or not item_id:
        raise ValueError(f"refresh_stock: invalid item_id: {item_id!r}")
    if not isinstance(stock, int) or isinstance(stock, bool):
        raise ValueError(
            f"refresh_stock: stock must be int, got {type(stock).__name__}"
        )
    if stock < 0:
        raise ValueError(f"refresh_stock: stock must be non-negative, got {stock}")
    if not isinstance(today_kst, str) or not today_kst:
        raise ValueError(f"refresh_stock: invalid today_kst: {today_kst!r}")

    _stock_col().update_one(
        {"itemId": item_id},
        {
            "$set": {"stock": stock, "lastRefresh": today_kst},
            "$setOnInsert": {"itemId": item_id},
        },
        upsert=True,
    )


def ensure_stock_entry(
    item_id: str, today_kst: str, default_stock: int
) -> dict:
    """문서 없으면 default_stock 으로 생성, 있으면 그대로 반환 (멱등)."""
    if not isinstance(item_id, str) or not item_id:
        raise ValueError(f"ensure_stock_entry: invalid item_id: {item_id!r}")
    if not isinstance(today_kst, str) or not today_kst:
        raise ValueError(
            f"ensure_stock_entry: invalid today_kst: {today_kst!r}"
        )
    if not isinstance(default_stock, int) or isinstance(default_stock, bool):
        raise ValueError(
            f"ensure_stock_entry: default_stock must be int, got {type(default_stock).__name__}"
        )
    if default_stock < 0:
        raise ValueError(
            f"ensure_stock_entry: default_stock must be non-negative, got {default_stock}"
        )

    col = _stock_col()
    existing = col.find_one({"itemId": item_id})
    if existing is not None:
        return existing

    doc = {
        "itemId": item_id,
        "stock": default_stock,
        "lastRefresh": today_kst,
    }
    result = col.insert_one(doc)
    doc["_id"] = result.inserted_id
    return doc


def get_stock(item_id: str) -> Optional[dict]:
    if not isinstance(item_id, str) or not item_id:
        raise ValueError(f"get_stock: invalid item_id: {item_id!r}")
    return _stock_col().find_one({"itemId": item_id})


def reduce_stock(item_id: str, qty: int) -> bool:
    """재고 atomic 차감. 부족 시 False."""
    if not isinstance(item_id, str) or not item_id:
        raise ValueError(f"reduce_stock: invalid item_id: {item_id!r}")
    if not isinstance(qty, int) or isinstance(qty, bool):
        raise ValueError(
            f"reduce_stock: qty must be int, got {type(qty).__name__}"
        )
    if qty <= 0:
        raise ValueError(f"reduce_stock: qty must be positive, got {qty}")

    result = _stock_col().update_one(
        {"itemId": item_id, "stock": {"$gte": qty}},
        {"$inc": {"stock": -qty}},
    )
    return result.modified_count > 0


def get_all_daily_stocks() -> list[dict]:
    return list(_stock_col().find().sort("itemId", 1))

File: packages/shared-db/src/crud/shop.ts
/**
 * shop_inventory + shop_daily_stock CRUD
 *
 * tia_bot 의 편의점 도메인.
 * - shop_inventory: 사용자(User) 단위 보유 소모품 (character_inventory 와 분리).
 * - shop_daily_stock: 아이템별 당일 재고. lastRefresh 는 KST YYYY-MM-DD 문자열.
 */

import type {
  ShopDailyStock,
  ShopInventory,
} from "../types/index.js";

import {
  shopDailyStockCol,
  shopInventoryCol,
} from "../collections.js";

/* ── shop_inventory ── */

/**
 * 활성 보유 인벤토리 (quantity > 0).
 * 일반 표시/조회용. 0 quantity 잔여 row 는 제외.
 */
export async function getUserInventory(userId: string): Promise<ShopInventory[]> {
  const col = await shopInventoryCol();
  return col
    .find({ userId, quantity: { $gt: 0 } })
    .sort({ itemId: 1 })
    .toArray();
}

/**
 * 0 quantity 포함 전체 인벤토리 (감사/마이그용).
 * race-aware delete 직전 또는 마이그레이션 검증 등 raw 데이터가 필요할 때만 사용.
 */
export async function getUserInventoryRaw(userId: string): Promise<ShopInventory[]> {
  const col = await shopInventoryCol();
  return col.find({ userId }).sort({ itemId: 1 }).toArray();
}

/**
 * 사용자 인벤토리에 아이템 qty 를 더한다 (upsert + $inc).
 * qty 는 양수.
 */
export async function addInventory(
  userId: string,
  itemId: string,
  qty: number,
): Promise<void> {
  if (qty <= 0) {
    throw new Error(`addInventory: qty must be positive, got ${qty}`);
  }
  const col = await shopInventoryCol();
  await col.updateOne(
    { userId, itemId },
    {
      $inc: { quantity: qty },
      $set: { updatedAt: new Date() },
      $setOnInsert: { userId, itemId },
    },
    { upsert: true },
  );
}

/**
 * 사용자 인벤토리에서 아이템 qty 를 atomic 하게 차감.
 *
 * - quantity >= qty 일 때만 매치하여 race condition 방지.
 * - 보유 부족 시 false 반환 (수정 없음).
 * - 차감 후 quantity 가 0 이 되면 race-aware deleteOne 으로 row 정리.
 */
export async function removeInventory(
  userId: string,
  itemId: string,
  qty: number,
): Promise<boolean> {
  if (qty <= 0) {
    throw new Error(`removeInventory: qty must be positive, got ${qty}`);
  }
  const col = await shopInventoryCol();
  const result = await col.findOneAndUpdate(
    { userId, itemId, quantity: { $gte: qty } },
    {
      $inc: { quantity: -qty },
      $set: { updatedAt: new Date() },
    },
    { returnDocument: "after" },
  );
  if (!result) return false;
  if (result.quantity === 0) {
    // race-aware delete: 다른 호출이 그 사이 +qty 했으면 quantity:0 매치 안 되어 보존됨.
    await col.deleteOne({ _id: result._id, quantity: 0 });
  }
  return true;
}

/* ── shop_daily_stock ── */

/**
 * lastRefresh 가 todayKst 와 다르면 true (재고 리프레시 필요).
 * 문서 미존재 시에도 true.
 */
export async function needsRefresh(
  itemId: string,
  todayKst: string,
): Promise<boolean> {
  const col = await shopDailyStockCol();
  const doc = await col.findOne({ itemId });
  if (!doc) return true;
  return doc.lastRefresh !== todayKst;
}

/**
 * 재고를 todayKst 기준으로 강제 갱신 (upsert).
 */
export async function refreshStock(
  itemId: string,
  stock: number,
  todayKst: string,
): Promise<void> {
  const col = await shopDailyStockCol();
  await col.updateOne(
    { itemId },
    {
      $set: { stock, lastRefresh: todayKst },
      $setOnInsert: { itemId },
    },
    { upsert: true },
  );
}

/**
 * 재고 문서가 없으면 defaultStock 으로 생성 후 반환 (멱등).
 * 이미 있으면 그대로 반환 (lastRefresh 갱신은 refreshStock 책임).
 */
export async function ensureStockEntry(
  itemId: string,
  todayKst: string,
  defaultStock: number,
): Promise<ShopDailyStock> {
  const col = await shopDailyStockCol();
  const existing = await col.findOne({ itemId });
  if (existing) return existing;

  const doc: ShopDailyStock = {
    itemId,
    stock: defaultStock,
    lastRefresh: todayKst,
  };
  const result = await col.insertOne(doc);
  return { ...doc, _id: result.insertedId };
}

export async function getStock(itemId: string): Promise<ShopDailyStock | null> {
  const col = await shopDailyStockCol();
  return col.findOne({ itemId });
}

/**
 * 재고를 atomic 하게 qty 차감.
 *
 * - stock >= qty 일 때만 매치.
 * - 재고 부족 시 false 반환.
 */
export async function reduceStock(
  itemId: string,
  qty: number,
): Promise<boolean> {
  if (qty <= 0) {
    throw new Error(`reduceStock: qty must be positive, got ${qty}`);
  }
  const col = await shopDailyStockCol();
  const result = await col.updateOne(
    { itemId, stock: { $gte: qty } },
    { $inc: { stock: -qty } },
  );
  return result.modifiedCount > 0;
}

export async function getAllDailyStocks(): Promise<ShopDailyStock[]> {
  const col = await shopDailyStockCol();
  return col.find().sort({ itemId: 1 }).toArray();
}
